### PR TITLE
:update: user details rebase

### PR DESCRIPTION
**Pull request by noob77777**
# User Login Flow Update
### Config: (assume we maintain the following configuration per user)
```
interface UserAddressConfig {
	id: int;
	alias: str;	// this is primary wallet address for non admin users
	registered_addresses: t.List[str]	// list of addresses the users has been verified for
}
```

### New Sign Up / Old User Sign In:

Instead of the current flow, LoginRequest will always take a list of addresses.

```
LoginRequest:
addresses: t.List[str]
```

- For new users, the first address is taken as default and we proceed with ergo-auth with the address at index 0.
- For mobile users the list should only have one address.
- For existing users we find out what their user id is from the list of addresses and the information we have in our database. We use the primary address already set to proceed with login.

### Change Primary Address:

```
PrimaryAddressChangeRequest:
	address: str
```

For primary address change, we first make sure that the user owns this address through ergoauth. Once verification is successful we add the new address to the registered_addresses list, and set the primary address to be the new one.
Only allow this for desktop.

Example:

`{ id: 1, alias: “addr1”, registered_addresses: [“addr1”] }`
to
`{ id: 1, alias: “addr2”, registered_addresses: [“addr1”, “addr2”] }`

## Changes to TokenHolders and GovernanceWhitelist:

To make sure address changes don’t cause issues to payments, we will store ergo_address_id from the ergo_addresses table instead of user_id in these tables.

## User Details API
- ### Adds CRUD API for user details
- ### Adds CRUD API for user notification/system settings

### Change Primary Address
```
@r.get(
    "/address_config/{user_id}",
    response_model=UserAddressConfig,
    response_model_exclude_none=True,
    name="user:user-address-config"
)
def user_address_config(
    user_id: int,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)

@r.post(
    "/change_primary_address",
    response_model=LoginRequestWebResponse,
    name="user:edit-user-address-config"
)
def edit_user_address_config(
    req: PrimaryAddressChangeRequest,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)

@r.post(
    "/verify_address/{request_id}",
    response_model=UserAddressConfig,
    name="user:verify-user-address-change"
)
def verify_user_address_change(
    request_id: str,
    authResponse: ErgoAuthResponse,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)
```

### User Profile Details
```
@r.get(
    "/details/{user_id}",
    response_model=UserDetails,
    response_model_exclude_none=True,
    name="users:user-details"
)
def user_details_all(
    user_id: int,
    db=Depends(get_db),
)

@r.get(
    "/profile/settings",
    response_model=UserProfileSettings,
    response_model_exclude_none=True,
    name="users:user-profile-settings"
)
def user_profile_settings(
    db=Depends(get_db),
    user=Depends(get_current_active_user),
)

@r.put(
    "/details/{user_id}", response_model=UserDetails, response_model_exclude_none=True, name="users:edit-details"
)
def edit_user_details(
    user_id: int,
    user_details: UpdateUserDetails,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)

@r.put(
    "/profile/settings", response_model=UserProfileSettings, response_model_exclude_none=True, name="users:edit-profile-settings"
)
def edit_user_settings(
    user_settings: UpdateUserProfileSettings,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)

@r.put(
    "/profile/follow", response_model_exclude_none=True, name="users:edit-profile-follow"
)
def user_profile_follow(
    req: FollowUserRequest,
    db=Depends(get_db),
    current_user=Depends(get_current_active_user),
)
```
## Rebases and Links
- https://github.com/ergo-pad/paideia-api/pull/6
- https://github.com/ergo-pad/ergopad-api/issues/414
- https://github.com/ergo-pad/ergopad-api/issues/375

### CODE REVIEW

File: app/api/auth.py [hunk 1]
```diff
@@ -30,20 +29,28 @@
 ##################################
 
 
-BASE_ERGOAUTH = "ergoauth://192.168.0.15:8000"
-BASE_URL = "http://192.168.0.15:8000"
+BASE_ERGOAUTH = "ergoauth://192.168.1.8:8000"
```
Comment (michaelmirandi) [inline app/api/auth.py:32]: to test in local env, do I have to spin up a node?

File: app/api/auth.py [hunk 2]
```diff
@@ -30,20 +29,28 @@
...
+BASE_URL = "http://192.168.1.8:8000"
 
 
 @r.post("/login", response_model=LoginRequestWebResponse, name="ergoauth:login-web")
 async def ergoauth_login_web(
-    address: LoginRequest
+    addresses: LoginRequest,
```
Comment (michaelmirandi) [inline app/api/auth.py:38]: list of addresses as input from nautilus or other provider

File: app/api/users.py [hunk 3]
```diff
@@ -24,7 +41,7 @@
 async def users_list(
     response: Response,
     db=Depends(get_db),
-    current_user=Depends(get_current_active_user),
+    current_user=Depends(get_current_active_superuser),
```
Comment (michaelmirandi) [inline app/api/users.py:44]: Need to be able to grab all users for searchable addresses. should be any user.

File: app/api/users.py [hunk 4]
```diff
@@ -118,3 +131,194 @@ async def user_delete(
...
+    except Exception as e:
+        return JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content=f'{str(e)}')
+
+
+@r.post(
+    "/verify_address/{request_id}",
+    response_model=UserAddressConfig,
+    name="user:verify-user-address-change"
+)
+def verify_user_address_change(
```
Comment (michaelmirandi) [inline app/api/users.py:197]: same process for changing an address. Always ask for a signature.

File: app/api/users.py [hunk 5]
```diff
@@ -118,3 +131,194 @@ async def user_delete(
...
+    try:
+        return edit_user_profile_settings(db, current_user.id, user_settings)
+    except Exception as e:
+        return JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content=f'{str(e)}')
+
+
+@r.put(
+    "/profile/follow", response_model_exclude_none=True, name="users:edit-profile-follow"
+)
+def user_profile_follow(
```
Comment (michaelmirandi) [inline app/api/users.py:313]: Nice!

File: app/db/crud/proposals.py [hunk 6]
```diff
@@ -238,8 +239,8 @@ def delete_proposal_by_id(db: Session, id: int):
     if not proposal:
         return None
     # delete stuff
-    db.query(ProposalReference).filter(ProposalReference.referred_proposal_id ==
-                                       id or ProposalReference.referring_proposal_id == id).delete()
+    db.query(ProposalReference).filter(or_(ProposalReference.referred_proposal_id ==
```
Comment (michaelmirandi) [inline app/db/crud/proposals.py:242]: sql alchemy is powerful!

File: app/db/models/users.py [hunk 7]
```diff
@@ -12,12 +12,40 @@ class User(Base):
...
+
+class UserProfileSettings(Base):
+    __tablename__ = "user_profile_settings"
+
+    id = Column(Integer, primary_key=True, index=True)
+    user_id = Column(Integer, unique=True)
+    settings = Column(JSON)
+
+
 class ErgoAddress(Base):
```
Comment (michaelmirandi) [inline app/db/models/users.py:49]: These ergo addresses can also be vesting keys, nfts, or tokens, right??

File: app/db/models/users.py [hunk 8]
```diff
@@ -12,12 +12,40 @@ class User(Base):
...
+    follower_id = Column(Integer)
+    followee_id = Column(Integer)
+
+
+class UserProfileSettings(Base):
+    __tablename__ = "user_profile_settings"
+
+    id = Column(Integer, primary_key=True, index=True)
+    user_id = Column(Integer, unique=True)
+    settings = Column(JSON)
```
Comment (michaelmirandi) [inline app/db/models/users.py:46]: Cool JSON storage.

File: app/api/auth.py (hunk 1 above)
Comment (noob77777) [reply app/api/auth.py:32]: no... you have to replace it with our local api ip address

File: app/db/models/users.py (hunk 7 above)
Comment (noob77777) [reply app/db/models/users.py:49]: you mean `tokenIds`?

**Pull request merged**